src/Symfony/Component/Finder/Iterator/FilterIterator.php

```diff
@@ -18,9 +18,24 @@
  *
  * @author Alex Bogomazov
  */
-abstract class FilterIterator extends \FilterIterator
+abstract class FilterIterator extends \FilterIterator implements \RecursiveIterator
```

Comment:
memoization was not the culprit (it was simply useless). The issue is marking filter iterators as being RecursiveIterator while using them in a recursive way provides a broken behavior (it will exclude too much things)